### ISSUE TITLE
Does the example work on localhost:5000 ?

### ISSUE THREAD

**Issue by r-chris** [thedirtyfew/dash-keycloak#2]
I am trying to run the example dash app, but it seems to fail when trying to connect to `http://localhost:5000` during launch, because - well - it hasn't launched, yet. My understanding is that for some reason when calling `FlaskKeycloak.from_kc_oidc_json(server, redirect_uri="https://localhost:5000/", config_path=config_path)` during startup, it immediately tried to call the `redirect_uri`, but obviously can not connect to it, yet, because the application hasn't been launched, yet. Am I missing something or should that call maybe happen after the flask server is already running?

This is the error I am seeing:

```
Traceback (most recent call last):
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "./dash-keycloak/flask_keycloak/examples/dash_example.py", line 16, in <module>
    FlaskKeycloak.from_kc_oidc_json(server, redirect_uri="https://localhost:5000/", config_path=config_path)
  File "./dash-keycloak/flask_keycloak/core.py", line 140, in from_kc_oidc_json
    prefix_callback_path=prefix_callback_path)
  File "./dash-keycloak/flask_keycloak/core.py", line 104, in __init__
    app.wsgi_app = AuthMiddleWare(app.wsgi_app, auth_handler, redirect_uri, uri_whitelist, prefix_callback_path)
  File "./dash-keycloak/flask_keycloak/core.py", line 66, in __init__
    self.auth_uri = self.auth_handler.auth_url(self.callback_uri)
  File "./dash-keycloak/flask_keycloak/core.py", line 28, in auth_url
    return self.keycloak_openid.auth_url(callback_uri)
  File "./dash-keycloak/venv/lib/python3.7/site-packages/keycloak/keycloak_openid.py", line 166, in auth_url
    params_path = {"authorization-endpoint": self.well_know()['authorization_endpoint'],
  File "./dash-keycloak/venv/lib/python3.7/site-packages/keycloak/keycloak_openid.py", line 157, in well_know
    return raise_error_from_response(data_raw, KeycloakGetError)
  File "./dash-keycloak/venv/lib/python3.7/site-packages/keycloak/exceptions.py", line 102, in raise_error_from_response
    response_body=response.content)
keycloak.exceptions.KeycloakGetError: 404: b'<html><head><title>Error</title></head><body>404 - Not Found</body></html>'
```

**Comment by emilhe** [comment]
On init, a call is made to the keycloak server to get some basic info. It seems like the app is unable to reach the keycloak server. If you open a browser, can you see the info? I think the url should be,

https://{your-keycloak-url}/auth/realms/{your-realm}/.well-known/openid-configuration

**Comment by r-chris** [comment]
Ok, so `redirect_uri="https://localhost:5000/"` should be pointing to my Keycloak server (e.g. `https://{your-keycloak-url}/auth/realms/{your-realm}/.well-known/openid-configuration`). I though it was the callback uri.

**Comment by emilhe** [comment]
No, the `redirect_uri` defines the page to which the user is redirected post login, i.e. typically the home page of your app. Were you able to access the keycloak url from your browser?

**Comment by r-chris** [comment]
Ok, that makes sense. Yes, I can log into Keycloak just fine - it's a publicly accessible DNS name. I can also access that endpoint you specified. I have a feeling that maybe my firewall prevents that service from contacting the url ...

**Comment by r-chris** [comment]
It's not the firewall - the same error still appears, when I disable my outgoing firewall :/

**Comment by emilhe** [comment]
You could try invoking the URL directly just before initializing Keycloak, i.e. something like

```
r = requests.get("https://{your-keycloak-url}/auth/realms/{your-realm}/.well-known/openid-configuration")
print(r.status_code)
```

If that works, the problem is not related to connectivity. If it does not work, something is blocking your python script.

**Comment by r-chris** [comment]
So, I did exactly what you suggested. Using `requests.get` with the full URL did work just fine. I got the expected return. However, I traced the error into the keycloak package. `keycloak_openid.json` on line 155 says this:

`data_raw = self.connection.raw_get(URL_WELL_KNOWN.format(**params_path))`

Somehow it concatenates the `base_url` and the `URL_WELL_KNOWN` without making sure that the `base_url` has a trailing slash. Therefore, the request it made internally went to: `"https://{your-keycloak-url}/authrealms/{your-realm}/.well-known/openid-configuration"` (missing `/` between `auth` and `realms`).

The config file I downloaded (`keycloak.json`) from my keycloak server, defined the `auth-server-url` without a trailing slash. I changed that in the `keycloak.json` file now, but it sounds like something that could be solved by using a smarter join (like `os.path.join` does for example).

**Comment by r-chris** [comment]
tl;dr - the problem was that the `auth-server-url` inside `keycloak_openid.json` needs to end with a trailing slash `/`.

**Comment by emilhe** [comment]
Ah, i see. I will take a look at it. But I guess you got it working then?

**Comment by r-chris** [comment]
Yes, it's working and I think the issue is in a dependency.
